### PR TITLE
Refactoring of https connector to make root certificates optional

### PR DESCRIPTION
**Pull request by pawelchcki**
Root certificates can be missing on the machine/container this library runs in.  Currently the existing hyper_rusttls implementation with panic, and exit. 

One way to avoid this is to rewrite part of the client initialization code to make it fail gracefully. Started this PR to see how that could look, but don't have time this week to bring this past the finish line

### CODE REVIEW

File: ddprof-exporter/src/connector/conn_stream.rs [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+use pin_project_lite::pin_project;
+
+pin_project! {
+    #[derive(Debug)]
+    #[project = ConnStreamProj]
+    pub enum ConnStream {
+        Tcp{ #[pin] transport: tokio::net::TcpStream },
+        Tls{ #[pin] transport: tokio_rustls::client::TlsStream<tokio::net::TcpStream>},
+        // Tokio doesn't handle unix sockets on windows
+        #[cfg(unix)]
```
[the remark this answers is not in the document]
Comment (paullegranddc) [reply ddprof-exporter/src/connector/conn_stream.rs:20]: Oh nice, I thought `pin_project_lite` didn't understand cfg macros

File: ddprof-exporter/src/connector/mod.rs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+use std::task::{Context, Poll};
+
+#[cfg(unix)]
+pub mod uds;
+
+mod conn_stream;
+use conn_stream::{ConnStream, ConnStreamError};
+
+#[derive(Clone)]
+pub enum MaybeHttpsConnector {
```
Comment (paullegranddc) [inline ddprof-exporter/src/connector/mod.rs:19]: Slight nitpick, because the connector also handles UDS even if it's not a member of the enum because we don't need to have any associated state, `MaybeHttpsConnector` is IMO not a great name.

**Review by paullegranddc — APPROVED** [review]

File: ddprof-exporter/src/connector/mod.rs (hunk 2 above)
Comment (pawelchcki) [reply ddprof-exporter/src/connector/mod.rs:19]: >MaybeHttpsConnector is IMO not a great name.
😅 
I was "inspired" by one of underlying classes -> MaybeHtttpsStream :)

I think I can rename it back now to just Connector (I used MaybeHttpsConnector name, when working on refactoring while I had old Connector still present.

**Comment by pawelchcki** [comment]
https://github.com/DataDog/libddprof/pull/47#discussion_r848363219

TBF I'm not certain it works as expected, I'll try quickly adding a Windows build 🤞

**Comment by paullegranddc** [comment]
Ok, I've tried compiling locally for windows using [cross](https://github.com/cross-rs/cross) , and it actually doesn't work.

I get this error

```
% cross check --target x86_64-pc-windows-gnu

error[E0412]: cannot find type `UnixStream` in module `tokio::net`
  --> ddprof-exporter/src/connector/conn_stream.rs:21:44
   |
21 |         Udp{ #[pin] transport: tokio::net::UnixStream },
   |                                            ^^^^^^^^^^ not found in `tokio::net`

error[E0599]: no variant named `Udp` found for enum `ConnStream`
  --> ddprof-exporter/src/connector/conn_stream.rs:21:9
   |
13 | / pin_project! {
14 | |     #[derive(Debug)]
15 | |     #[project = ConnStreamProj]
16 | |     pub enum ConnStream {
...  |
21 | |         Udp{ #[pin] transport: tokio::net::UnixStream },
   | |         ^^^ variant not found in `ConnStream`
22 | |     }
23 | | }
   | |_- variant `Udp` not found here
```

**Comment by paullegranddc** [comment]
Maybe if the cleaner code is worth it, we could try using `https://github.com/taiki-e/pin-project` instead of pin_project_lite as it supports more features.

**Comment by paullegranddc** [comment]
Or we just write the projection code ourselves. Looking at the kind of code generated by pin-project https://github.com/taiki-e/pin-project/blob/HEAD/examples/struct-default-expanded.rs , anf removing the cruft it doesn't look to bad.

**Comment by gleocadie** [comment]
> > [#47 (comment)](https://github.com/DataDog/libddprof/pull/47#discussion_r848363219)
> > TBF I'm not certain it works as expected, I'll try quickly adding a Windows build 🤞
> 
> @gleocadie fyi, as you have interests in building on windows

thanks for the heads up

**Comment by pawelchcki** [comment]
Thanks @paullegranddc - with your suggestion I reimplemented the `project()` method. 
Its a bit verbose but not by much - explicit and visible use of Unsafe is a bit of an eyesore. 

https://github.com/DataDog/libddprof/pull/47/files#diff-1d9699d6f15d9c4af24ade58efdaadaa21f28f7150cdbb6a95f65c20e8eca6f9R42-R60

Hmm I'll switch back to pin_project (non-lite) just to reduce the number of visible code lines.

File: ddprof-exporter/src/connector/conn_stream.rs [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+#[derive(Debug)]
+#[pin_project(project=ConnStreamProj)]
+pub enum ConnStream {
+    Tcp {
+        #[pin]
+        transport: tokio::net::TcpStream,
+    },
+    Tls {
+        #[pin]
+        transport: Box<tokio_rustls::client::TlsStream<tokio::net::TcpStream>>,
```
Comment (paullegranddc) [inline ddprof-exporter/src/connector/conn_stream.rs:22]: I think that the decision to box or not should be based on how likely we are to be in a situation where Https is not available.
If we’re likely to downgrade to http and use the Tcp variant, the size saving per connection may be meaningful if we open a lot of them (about 500 byte per connection. Are we going to that many, I doubt it though?)
If https is almost always available, we’re just adding an indirection to save space in a minority of cases

Honestly I would be in favor of just ignoring the lint and not boxing, but it probably doesn't matter that much

**Review by paullegranddc — APPROVED** [review]

File: ddprof-exporter/src/connector/conn_stream.rs (hunk 3 above)
Comment (pawelchcki) [reply ddprof-exporter/src/connector/conn_stream.rs:22]: Yeah I've had the same thoughts. 

I opted for boxing instead of ignoring the lint for 2 reasons:
1. Adding an indirection for a IO connection should not affect performance, and ignoring lints IMO is a bad precedent - unless there are clear benefits
2. It actually has some potential for space saving.

tl;dr; potentially single use box, looks to me eyes better than an ignored lint :)

Comment (paullegranddc) [reply ddprof-exporter/src/connector/conn_stream.rs:22]: I'm ok with Boxing

**Comment by pawelchcki** [comment]
Thanks All for the review!
